### ISSUE TITLE
utility: kraken-style LCA classification on banded signatures

### ISSUE THREAD

**Issue by ctb** [dib-lab/sourmash#302]
See https://gist.github.com/ctb/9deb40a68108256ab4fd84c6b8e92e01 for implementation.

cc @brooksph @taylorreiter

**Comment by ctb** [comment]
Some more thoughts:

* scaling Kraken database building and search by a factor of 100x would already be pretty amazing;
* we should evaluate Bracken on banded output; that would be the major test
* support FASTA input in addition to signatures
* we should support fast merging/updating/etc of the databases.
* supporting multiple k-mer sizes is also a great idea!

**Comment by ctb** [comment]
* propose to add as `sourmash lca`
* support multiple query databases? would need nodes.dmp.
* deliver/store as .json file with --scaled and ksize values for multiple lca k-mer size dbs
* kraken compatible output possible?

**Comment by ctb** [comment]
Current output:
```
% kraken/classify.py genbank/nodes.dmp genbank/names.dmp genbank-k31.lca sig-to-classify.sig
loading nodes_dmp / taxonomic tree
loading k-mer DB
found LCA classifications for 411 of 2137 hashes
19.23   411     0       U       131567  cellular organisms
18.16   388     23      U       2       cellular organisms;Bacteria
17.45   373     15      U       1783272 cellular organisms;Bacteria;Terrabacteria group
17.45   373     0       U       201174  cellular organisms;Bacteria;Terrabacteria group;Actinobacteria
17.13   366     7       U       1760    cellular organisms;Bacteria;Terrabacteria group;Actinobacteria;Actinobacteria
16.75   358     8       U       85007   cellular organisms;Bacteria;Terrabacteria group;Actinobacteria;Actinobacteria;Corynebacteriales
16.75   358     0       U       1762    cellular organisms;Bacteria;Terrabacteria group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae
15.58   333     25      U       1763    cellular organisms;Bacteria;Terrabacteria group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae;Mycobacterium
0.19    4       329     U       77643   cellular organisms;Bacteria;Terrabacteria group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae;Mycobacterium;Mycobacterium tuberculosis complex
```

LCA files for genbank-k21, k31, and k51 are available [on the OSF](https://osf.io/vk4fa/) under `sourmash-lca-mark1`.

They were built with the command

```
python gist/extract.py genbank*.csv.gz nodes.dmp --traverse-directory .sbt.genba
nk-k21/ --savename genbank-k21.lca -k 21 --scaled 10000
```

and each took approximately 2 hours and 6 GB of RAM to build on the MSU HPCC;

```
lca.o46877939:    resources_used.walltime = 02:51:58
lca.o46877939:    resources_used.vmem = 6294516kb
```

**Comment by ctb** [comment]
More TODO items:

* add to sourmash as `sourmash lca search` and `sourmash lca index` (?), w/tests etc.
* support a short-read search mode.

**Comment by ctb** [comment]
Random other thoughts:

* nodes/names should be distributed with LCA databases, maybe;
* do caching of nodes/names?
* LCA databases are small enough to distribute all together, at least for scaled=10k;
* provide JSON file linking LCA databases, ksize, nodes/name files;
* maybe support sqlite or leveldb databases? and/or redis?

**Comment by ctb** [comment]
Fixed in #367.
